### PR TITLE
GH-41343: [C++][CMake] Remove unused ARROW_NO_DEPRECATED_API

### PR DESCRIPTION
**Pull request by wgtmac**
### Rationale for this change

ARROW_NO_DEPRECATED_API is not used in the source code.

### What changes are included in this PR?

Remove ARROW_NO_DEPRECATED_API cmake variable.

### Are these changes tested?

Pass CIs.

### Are there any user-facing changes?

Perhaps yes, users used to set ARROW_NO_DEPRECATED_API to build Arrow will see a warning for unknown CMake variable.
* GitHub Issue: #41343

### CODE REVIEW

File: docs/source/developers/cpp/building.rst
Comment (kou) [inline docs/source/developers/cpp/building.rst:632]: Can we mention the `-Werror=deprecated-declarations` option instead of removing this section entirely?

**Review by kou — APPROVED** [review]
+1

But I want to hear opinions from others.

File: docs/source/developers/cpp/building.rst
Comment (wgtmac) [reply docs/source/developers/cpp/building.rst:632]: It seems we need to mention different kinds of compilers as the flag varies.

Comment (wgtmac) [reply docs/source/developers/cpp/building.rst:632]: https://github.com/apache/arrow/blob/299eb26e8c22b4aad4876c9e3b52f9adde699a5c/cpp/src/arrow/util/macros.h#L136

Is the definition of `__COVERITY__` still used? I cannot find it anywhere else. @kou

Comment (kou) [reply docs/source/developers/cpp/building.rst]: @pitrou Are we still use Coverity?

File: docs/source/developers/cpp/building.rst
Comment (wgtmac) [inline docs/source/developers/cpp/building.rst:630]: I've rephrased here. WDYT? @kou

File: docs/source/developers/cpp/building.rst
Comment (pitrou) [reply docs/source/developers/cpp/building.rst]: No. We've only had once access to a Coverity report of Arrow C++, a long time ago. It actually didn't surface any major issue.

**Comment by wgtmac** [comment]
I will merge this by the end of this week if there is no objection.

**Comment by wgtmac** [comment]
Rebased on the latest main branch and wait for CIs before merging.

**Comment by wgtmac** [comment]
All CI failures are unrelated. But I have found the following failure from `C++ / ARM64 Ubuntu 20.04 C++`. Let me take a look and fix it in a separate issue if necessary.

```
66/94 Test #69: arrow-dataset-file-parquet-encryption-test ...***Failed    0.26 sec
Running arrow-dataset-file-parquet-encryption-test, redirecting output into /build/cpp/build/test-logs/arrow-dataset-file-parquet-encryption-test.txt (attempt 1/1)
/arrow/cpp/build-support/run-test.sh: line 88: 35257 Segmentation fault      (core dumped) $TEST_EXECUTABLE "$@" > $LOGFILE.raw 2>&1
Running main() from _deps/googletest-src/googletest/src/gtest_main.cc
[==========] Running 3 tests from 2 test suites.
[----------] Global test environment set-up.
[----------] 2 tests from DatasetEncryptionTest
[ RUN      ] DatasetEncryptionTest.WriteReadDatasetWithEncryption
/build/cpp/src/arrow/dataset
```

**Pull request merged**
